modules/diskq/qdisk.c

```diff
@@ -1355,10 +1297,65 @@ qdisk_free(QDisk *self)
   g_free(self);
 }
 
+gboolean
+qdisk_start_serialize(QDisk *self, gpointer *args)
+{
+  if (self->start_serialize)
+    return self->start_serialize(args);
+
+  return FALSE;
+}
+
+gboolean
+qdisk_end_serialize(QDisk *self, gpointer *args)
+{
+  if (self->end_serialize)
+    return self->end_serialize(args);
+
+  return FALSE;
+}
+
+
+static gboolean
+_start_serialize(gpointer *args)
+{
+  SerializeArchive *sa = args[0];
+  GError **error = args[1];
+
+  /* Leave space for the real record_length for later */
+  if (!serialize_write_uint32(sa, 0))
+    {
+      g_set_error(error, G_FILE_ERROR, G_FILE_ERROR_IO, "cannot write record length");
+      return FALSE;
+    }
+
+  return TRUE;
+
+}
+
+static gboolean
+_end_serialize(gpointer *args)
+{
+  GString *serialized = args[0];
+  GError **error = args[1];
+
+  guint32 record_length = GUINT32_TO_BE(serialized->len - sizeof(guint32));
+  if (record_length == 0)
+    {
+      g_set_error(error, G_FILE_ERROR, G_FILE_ERROR_IO, "message is empty");
+      return FALSE;
+    }
+
+  g_string_overwrite_len(serialized, 0, (gchar *) &record_length, sizeof(guint32));
+  return TRUE;
+}
+
 QDisk *
 qdisk_new(DiskQueueOptions *options, const gchar *file_id)
 {
   QDisk *self = g_new0(QDisk, 1);
   qdisk_init_instance(self, options, file_id);
+  self->start_serialize = _start_serialize;
+  self->end_serialize = _end_serialize;
```

Comment:
There is no reason for these 2 to be virtual. They are set here, they are not meant to be overridden.

Reply:
In the future, if we want to create different methods of saving data to file (qdisk), then we have the options to create. And that's why the args argumentum.

But if we agree to we want only this method, then I will remove the virtual function (and the args parameter)

Reply:
It's `QDisk` responsibility to define how to store a binary record. Even if there was a different method,  that would be a private implementation detail (but I don't think another method could be expected in the foreseeable future).

Reply:
If we remove the responsibility of managing this in log_queue_disk_serialize() function and just do this where this was called, we would still have the options for changing the on-disk format if we ever have to. But I don't anticipate that at this point either.